### PR TITLE
feat: add dump for consul_kv

### PR DESCRIPTION
**Pull request by yongboy**
### What this PR does / why we need it:

1. Add dump support for `consul_kv`, more intro:  https://github.com/yongboy/apisix/blob/consul_dump/docs/en/latest/discovery/consul_kv.md#dump_data
2. Avoid hard code discovery module's private control api fully URI 

Define the module's control api short path, eg `/show_dump_file`

```lua
function _M.control_api()
    return {
        {
            methods = {"GET"},
            uris = {"/show_dump_file"},
            handler = show_dump_file,
        }
    }
end
```

And, when you access it,  you need to type it's fully URI:

```shell
GET /v1/discovery/{Module_Name}/show_dump_file
```

eg: 

```shell
GET /v1/discovery/consul_kv/show_dump_file
```

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: apisix/discovery/consul_kv.lua [hunk 1]
```diff
@@ -356,6 +434,20 @@ function _M.init_worker()
         return
     end
 
+    if consul_conf.dump then
+      local dump = consul_conf.dump
+      if not dump.path or #dump.path == 0 then
```
Comment (spacewander) [inline apisix/discovery/consul_kv.lua:439]: We can use `required` & `minLength` to validate the path in the schema.

File: apisix/discovery/consul_kv.lua [hunk 2]
```diff
@@ -230,6 +240,70 @@ local function update_application(server_name_prefix, data)
...
+    local succ, err =  util.write_file(dump_params.path, data)
+    if not succ then
+        core.log.error("write dump into file got error: ", err)
+    end
+end
+
+
+local function show_dump_file()
+    if not dump_params then
+        return 200, "dump params is nil"
```
Comment (spacewander) [inline apisix/discovery/consul_kv.lua:295]: Should we use a non-200 status code when an error happened?
And I think we need to test for error in this function.

File: t/discovery/consul_kv_dump.t [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+            local t = require("lib.test_admin")
+            ngx.sleep(2)
+
+            local code, body, res = t.test('/v1/discovery/consul_kv/show_dump_file',
+                ngx.HTTP_GET)
+            local entity = json.decode(res)
+            ngx.say(json.encode(entity.services))
+        }
+    }
+--- timeout: 12
```
Comment (spacewander) [inline t/discovery/consul_kv_dump.t:100]: Do we need so long timeout?

File: apisix/control/router.lua [hunk 4]
```diff
@@ -31,6 +31,41 @@ local get_method = ngx.req.get_method
...
+    if not core.string.has_prefix(uri, "/") then
+        core.table.insert(tmp, "/")
+    end
+    core.table.insert(tmp, uri)
+
+    return core.table.concat(tmp, "")
+end
+
+-- we do not hardcode the discovery module's control api uri
+local function format_dismod_controlapi_uris(mod_name, api_route)
```
Comment (spacewander) [inline apisix/control/router.lua:49]: Better to use `format_dismod_control_api_uris`

File: docs/en/latest/discovery/consul_kv.md [hunk 5]
```diff
@@ -73,6 +76,30 @@ The `keepalive` has two optional values:
...
+
+The `dump` has three optional values now:
+
+- `path`, the dump file save path
+    - support relative path, eg: `logs/consul_kv.dump`
+    - support absolute path, eg: `/tmp/consul_kv.bin`
+    - make sure the dump file's parent path exist
+    - make sure the `apisix` has the dump file's read-write access permission,eg: `chown  www:root conf/upstream.d/`
+- `load_on_init`, default value is `true`
+    - if `true`, load the data from dump file before load data from consul
```
Comment (spacewander) [inline docs/en/latest/discovery/consul_kv.md:97]: Look like we need to prepare a dump file for a fresh installed APISIX? Maybe we need to write down this in the docs?

File: apisix/discovery/consul_kv.lua [hunk 6]
```diff
@@ -319,7 +397,7 @@ local function format_consul_params(consul_conf)
         if scheme ~= "http" then
             return nil, "only support consul http schema address, eg: http://address:port"
         elseif path ~= "/" or core.string.has_suffix(v, '/') then
-            return nil, "invalid consul server address, the valid format: http://address:port"
+            return nil, "invald consul server address, the valid format: http://address:port"
```
[the remark this answers is not in the document]
Comment (yongboy) [reply apisix/discovery/consul_kv.lua:400]: Y

[the remark this answers is not in the document]
Comment (yongboy) [reply apisix/discovery/consul_kv.lua:400]: Y

File: apisix/discovery/consul_kv.lua (hunk 1 above)
Comment (yongboy) [reply apisix/discovery/consul_kv.lua:439]: Y

File: apisix/discovery/consul_kv.lua (hunk 2 above)
Comment (yongboy) [reply apisix/discovery/consul_kv.lua:295]: 1. You are right, we should return 500 status code when an error happened
2. Add Test for `show_dump_file` function when get an error ?

File: t/discovery/consul_kv_dump.t (hunk 3 above)
Comment (yongboy) [reply t/discovery/consul_kv_dump.t:100]: Yes, it should be 3 sec better :))

File: apisix/control/router.lua (hunk 4 above)
Comment (yongboy) [reply apisix/control/router.lua:49]: Y

File: docs/en/latest/discovery/consul_kv.md (hunk 5 above)
Comment (yongboy) [reply docs/en/latest/discovery/consul_kv.md:97]: Y

File: apisix/discovery/consul_kv.lua (hunk 2 above)
Comment (spacewander) [reply apisix/discovery/consul_kv.lua:295]: > Add Test for show_dump_file function when get an error ?

Yes

File: apisix/discovery/consul_kv.lua [hunk 7]
```diff
@@ -230,6 +241,73 @@ local function update_application(server_name_prefix, data)
...
+    local succ, err =  util.write_file(dump_params.path, data)
+    if not succ then
+        core.log.error("write dump into file got error: ", err)
+    end
+end
+
+
+local function show_dump_file()
+    if not dump_params then
+        return 500, "dump params is nil"
```
Comment (spacewander) [inline apisix/discovery/consul_kv.lua:299]: Personally speaking, I prefer to use 503 as 500 is for unexpected error.

File: docs/en/latest/discovery/consul_kv.md [hunk 8]
```diff
@@ -73,6 +76,31 @@ The `keepalive` has two optional values:
...
+
+- `path`, the dump file save path
+    - support relative path, eg: `logs/consul_kv.dump`
+    - support absolute path, eg: `/tmp/consul_kv.bin`
+    - make sure the dump file's parent path exist
+    - make sure the `apisix` has the dump file's read-write access permission,eg: `chown  www:root conf/upstream.d/`
+- `load_on_init`, default value is `true`
+    - if `true`, will try load the data from dump file before load data from consul
+    - if `false`, ignore load data
+    - and anyhow, we don't need to prepare a dump file for apisix at anytime
```
Comment (spacewander) [inline docs/en/latest/discovery/consul_kv.md:99]: This sentence confuses me. Does it look like we need to prepare a dump file if load_on_init is true before starting APISIX?

```
local function read_dump_srvs()
    local data, err = util.read_file(dump_params.path)
    if not data then
        log.warn("read dump file get error: ", err)
        return
    end
```

File: apisix/discovery/consul_kv.lua (hunk 7 above)
Comment (yongboy) [reply apisix/discovery/consul_kv.lua:299]: En, the 503 would be better here, for service unavailable, not ready to handle the request.

File: docs/en/latest/discovery/consul_kv.md (hunk 8 above)
Comment (yongboy) [reply docs/en/latest/discovery/consul_kv.md:99]: What about this ?

- `load_on_init`, default value is `true`
    - if `true`, just try to load the data from the dump file before loading data from    consul when starting
    - if `false`, ignore loading data from the dump file
    - Whether `true` or `false`, we don't need to prepare a dump file for apisix at anytime

Comment (spacewander) [reply docs/en/latest/discovery/consul_kv.md:99]: If `load_on_init` is true, we need to prepare a dump file, otherwise a warning will be logged.

Comment (yongboy) [reply docs/en/latest/discovery/consul_kv.md:99]: Here is my doc:

- `load_on_init`, default value is `true`
    - if `true`, just try to load the data from the dump file before loading data from  consul when starting, does not care the dump file exists or not
    - if `false`, ignore loading data from the dump file
    - Whether `true` or `false`, we don't need to prepare a dump file for apisix at anytime

File: apisix/discovery/consul_kv.lua [hunk 9]
```diff
@@ -230,6 +241,73 @@ local function update_application(server_name_prefix, data)
...
+
+
+local function show_dump_file()
+    if not dump_params then
+        return 503, "dump params is nil"
+    end
+
+    local data, err = util.read_file(dump_params.path)
+    if not data then
+        return 500, err
```
Comment (spacewander) [inline apisix/discovery/consul_kv.lua:304]: Should we use 503 for this too?

File: apisix/discovery/consul_kv.lua [hunk 10]
```diff
@@ -230,6 +241,73 @@ local function update_application(server_name_prefix, data)
 end
 
 
+local function read_dump_srvs()
+    local data, err = util.read_file(dump_params.path)
+    if not data then
+        log.warn("read dump file get error: ", err)
```
Comment (spacewander) [inline apisix/discovery/consul_kv.lua:247]: If we don't care about the existence of the dump file, maybe we can use a lower log level when we don't find the file?

File: apisix/discovery/consul_kv.lua (hunk 9 above)
Comment (yongboy) [reply apisix/discovery/consul_kv.lua:304]: Yes, it would be better.

File: apisix/discovery/consul_kv.lua (hunk 10 above)
Comment (yongboy) [reply apisix/discovery/consul_kv.lua:247]: No, haha :))
Use warn level is better than info level here for unexpected event.

Comment (spacewander) [reply apisix/discovery/consul_kv.lua:247]: So you suggest to prepare the dump file before staring APISIX to avoid the unexpected event (file not found)?

Comment (yongboy) [reply apisix/discovery/consul_kv.lua:247]: No, we don't prepare the dump file.
Maybe the log level is `notice` would be suitable.

File: apisix/discovery/consul_kv.lua [hunk 11]
```diff
@@ -244,7 +244,7 @@ end
 local function read_dump_srvs()
     local data, err = util.read_file(dump_params.path)
     if not data then
-        log.warn("read dump file get error: ", err)
+        log.notice("read dump file get error: ", err)
```
Comment (spacewander) [inline apisix/discovery/consul_kv.lua:247]: I think we can check if the file exists first via 
https://refined-github-html-preview.kidonng.workers.dev/lunarmodules/Penlight/raw/master/docs/libraries/pl.path.html#exists

If not exists, log a notice message.

And then try to read it.

**Review by spacewander — APPROVED** [review]

File: apisix/discovery/consul_kv.lua (hunk 11 above)
Comment (yongboy) [reply apisix/discovery/consul_kv.lua:247]: Y

Comment (spacewander) [reply apisix/discovery/consul_kv.lua:247]: @yongboy 
We just add the heavy library in the master branch: https://github.com/apache/apisix/blob/1064e2d4b7edcc6f590ca69dc34a100ae6bbc12c/rockspec/apisix-master-0.rockspec#L68.
Would be fine if we can find another usage of this library.

Comment (yongboy) [reply apisix/discovery/consul_kv.lua:247]: I think we don't need toimport the heavy library `penlight` at the moment. It is so heavy!
If another one needs it really, then we can change the code later.

**Comment by yongboy** [comment]
@spacewander 
How to fix the ci error, maybe you can help me :))

```
#   Failed test 'ERROR: client socket timed out - TEST 9: disable search option
# '
#   at /home/runner/work/apisix/apisix/test-nginx/lib/Test/Nginx/Socket.pm line 2094.

#   Failed test 'TEST 9: disable search option - status code ok'
#   at /home/runner/work/apisix/apisix/test-nginx/lib/Test/Nginx/Socket.pm line 948.
#          got: ''
#     expected: '200'

#   Failed test 'TEST 9: disable search option - pattern "error: failed to query the DNS server" should match a line in error.log (req 0)'
#   at /home/runner/work/apisix/apisix/test-nginx/lib/Test/Nginx/Socket.pm line 1213.
t/core/utils.t ............................. 
Failed 3/25 subtests 
```

**Comment by spacewander** [comment]
@yongboy 
You can merge master:
https://github.com/apache/apisix/commit/7ae302034a97da1ba33fd136f7892ba44f33abbf

**Review by tokers — APPROVED** [review]

**Pull request merged**